File: ide/macos/Tests/IDECoreTests/SemanticTokenDecoderTests.swift
import XCTest
@testable import IDECore

final class SemanticTokenDecoderTests: XCTestCase {
    func testDecodesHandComputedDeltaExample() {
        // line 0: 'module' char 0 len 6, keyword (index 9)
        // line 2: 'model'  char 0 len 5, keyword
        // line 2: 'Greeting' char 6 len 8, model (index 14), declaration (bit 0)
        let data = [
            0, 0, 6, 9, 0,
            2, 0, 5, 9, 0,
            0, 6, 8, 14, 1,
        ]

        let spans = SemanticTokenDecoder.decode(data)

        XCTAssertEqual(spans.count, 3)
        XCTAssertEqual(spans[0], TokenSpan(line: 0, startCharacter: 0, length: 6, typeName: "keyword", modifiers: []))
        XCTAssertEqual(spans[1].line, 2)
        XCTAssertEqual(spans[1].startCharacter, 0)
        XCTAssertEqual(spans[2].line, 2)
        XCTAssertEqual(spans[2].startCharacter, 6)
        XCTAssertEqual(spans[2].typeName, "model")
        XCTAssertEqual(spans[2].modifiers, ["declaration"])
    }

    func testEmptyDataDecodesToNothing() {
        XCTAssertTrue(SemanticTokenDecoder.decode([]).isEmpty)
    }

    func testLegendMatchesDocs21Order() {
        XCTAssertEqual(SemanticTokenDecoder.tokenTypes.count, 21)
        XCTAssertEqual(SemanticTokenDecoder.tokenTypes[9], "keyword")
        XCTAssertEqual(SemanticTokenDecoder.tokenTypes[14], "model")
        XCTAssertEqual(SemanticTokenDecoder.tokenTypes[20], "nativeTarget")
    }
}

File: ide/windows/tests/ClearKryptIDE.Core.Tests/SemanticTokenDecoderTests.cs
using ClearKryptIDE.Core.Lsp;
using Xunit;

namespace ClearKryptIDE.Core.Tests;

public sealed class SemanticTokenDecoderTests
{
    [Fact]
    public void DecodesDeltaEncodedTokensAgainstTheLegend()
    {
        // Hand-computed for:
        //   line 0: 'module' at char 0, length 6, keyword (index 9), no modifiers
        //   line 2: 'model' at char 0, length 5, keyword
        //   line 2: 'Greeting' at char 6, length 8, model (index 14), declaration (bit 0)
        var data = new[]
        {
            0, 0, 6, 9, 0,
            2, 0, 5, 9, 0,
            0, 6, 8, 14, 1,
        };

        var spans = SemanticTokenDecoder.Decode(data);

        Assert.Equal(3, spans.Count);
        Assert.Equal(new SemanticTokenSpan(0, 0, 6, "keyword", Array.Empty<string>()), spans[0] with { Modifiers = Array.Empty<string>() });
        Assert.Equal(0, spans[0].Line);
        Assert.Equal("keyword", spans[0].TokenType);

        Assert.Equal(2, spans[1].Line);
        Assert.Equal(0, spans[1].StartCharacter);

        Assert.Equal(2, spans[2].Line);
        Assert.Equal(6, spans[2].StartCharacter);
        Assert.Equal(8, spans[2].Length);
        Assert.Equal("model", spans[2].TokenType);
        Assert.Contains("declaration", spans[2].Modifiers);
    }

    [Fact]
    public void EmptyDataDecodesToNoTokens()
    {
        Assert.Empty(SemanticTokenDecoder.Decode(Array.Empty<int>()));
    }
}
